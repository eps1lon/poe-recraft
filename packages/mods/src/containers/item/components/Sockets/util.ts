import Item from '../../Item';
import { Socket, SocketColor } from './types';
import ValueRange from '../../../../calculator/ValueRange';

/**
 * creates a Socket according to the highest attr requirement
 * if two or more requirements are equal the color is chosen
 * with the priority Green > Red > Blue
 * @param item
 */
export function rollRequirementBiasedStable(item: Item): Socket {
  const { dex, int, str } = item.requirements.list();
  const color_by_attr = [
    { value: dex.value, color: SocketColor.green },
    { value: str.value, color: SocketColor.red },
    { value: int.value, color: SocketColor.blue },
    // descending
  ].sort((a, b) => -ValueRange.cmpMax(a.value, b.value));

  return {
    color: color_by_attr[0].color,
  };
}
