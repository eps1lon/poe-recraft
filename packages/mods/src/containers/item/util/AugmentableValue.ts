/**
 * a numeric value that was augmented in some way
 *
 * {augmented} is true if the the base value was changed to {value}
 */
export default interface AugmentableValue {
  value: number | [number, number];
  augmented: boolean;
}

export function isZero(
  augmentable: { value: AugmentableValue['value'] } | null | undefined,
): boolean {
  if (augmentable == null) {
    return true;
  }
  const { value } = augmentable;

  return (
    value === 0 || (Array.isArray(value) && value[0] === 0 && value[1] === 0)
  );
}

export function anyCmp(
  augmentable: { value: AugmentableValue['value'] },
  cmp: (n: number) => boolean,
): boolean {
  const { value } = augmentable;

  if (typeof value === 'number') {
    return cmp(value);
  } else {
    return value.some(cmp);
  }
}
