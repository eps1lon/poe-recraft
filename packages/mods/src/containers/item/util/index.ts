export { default as AugmentableValue } from './AugmentableValue';
