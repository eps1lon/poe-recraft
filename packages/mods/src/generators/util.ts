import { GeneratorDetails } from './Generator';

// filter nothing
const allowAll = () => true;

export interface WithSpawnchance {
  spawnchance: number;
}

export const calculateSpawnchance = (
  collection: Array<GeneratorDetails<any>>,
  filter: (details: GeneratorDetails<any>) => boolean = allowAll,
): Array<GeneratorDetails<any> & WithSpawnchance> => {
  const sum_spawnweight = collection.reduce((sum, details) => {
    if (details.spawnweight != null && filter(details)) {
      return sum + details.spawnweight;
    } else {
      return sum;
    }
  }, 0);

  return collection.map(details => {
    let spawnchance = 0;

    if (details.spawnweight != null && filter(details)) {
      spawnchance = details.spawnweight / sum_spawnweight;
    }

    return {
      ...details,
      spawnchance,
    };
  });
};
