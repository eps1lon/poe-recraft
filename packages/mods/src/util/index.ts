export { Flags, anySet } from './Flags';
export { default as MetaData } from './MetaData';
