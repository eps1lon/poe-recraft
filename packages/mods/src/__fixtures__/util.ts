import * as fs from 'fs';
import * as path from 'path';

import {
  AtlasNodeProps,
  BaseItemTypeProps,
  CraftingBenchOptionsProps,
  EssenceProps,
  ModProps,
} from '../schema';

import AtlasNode from '../containers/AtlasNode';
import Item from '../containers/item/Item';
import { Essence, MasterBenchOption } from '../generators';
import Mod from '../mods/Mod';
import PropsTable, { TableProps } from '../helpers/PropsTable';
import { Buildable } from '../interfaces/Buildable';

const json_cache: Partial<Tables> = {};

export const createTable = <P extends TableProps, T, A1>(
  file: keyof Tables,
  constructor: Buildable<P, T, A1>,
  arg1: A1,
) => {
  if (json_cache[file] == null) {
    const body = fs.readFileSync(
      path.join(__dirname, `../../data/${file}.json`),
    );

    const json = JSON.parse(body.toString());

    // @ts-ignore
    json_cache[file] = new PropsTable(json, constructor, arg1);
  }

  return json_cache[file];
};

interface Tables {
  atlas: PropsTable<AtlasNodeProps, AtlasNode, undefined>;
  craftingbenchoptions: PropsTable<
    CraftingBenchOptionsProps,
    MasterBenchOption,
    undefined
  >;
  essences: PropsTable<EssenceProps, Essence, ModProps>;
  items: PropsTable<BaseItemTypeProps, Item, undefined>;
  mods: PropsTable<ModProps, Mod, undefined>;
}

export const createTables = () => {
  const tables = {
    atlas: createTable('atlas', AtlasNode, undefined),
    craftingbenchoptions: createTable(
      'craftingbenchoptions',
      MasterBenchOption,
      undefined,
    ),
    items: createTable('items', Item, undefined),
    mods: createTable('mods', Mod, undefined),
  } as Tables;

  // @ts-ignore: its not undefined though?!
  tables.essences = createTable('essences', Essence, tables.mods.all());

  return tables;
};

export const byWorldAreaId = (id: string) => (node: AtlasNodeProps) =>
  node.world_area.id === id;

export const fromWorldAreaId = (
  id: string,
  atlas: PropsTable<AtlasNodeProps, AtlasNode, undefined>,
): AtlasNode => {
  return atlas.from(node => node.world_area.id === id);
};

export const byEssenceName = (name: string) => (essence: EssenceProps) =>
  essence.base_item_type.name === name;
