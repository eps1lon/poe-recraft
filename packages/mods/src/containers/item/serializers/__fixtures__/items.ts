import { createTables } from '../../../../__fixtures__/util';

const { items, mods } = createTables();

export default [
  items
    .fromName('Platinum Kris')
    .rarity.set('magic')
    .addMod(
      mods.fromId('LocalIncreasedPhysicalDamagePercentAndAccuracyRating2'),
    )
    .setProperty('mirrored', true),
  items
    .fromName('Hubris Circlet')
    .rarity.set('rare')
    .asShaperItem()
    .addMod(mods.fromId('IncreasedLife0'))
    .addMod(mods.fromId('LocalIncreasedEnergyShield2'))
    .addMod(mods.fromId('LocalIncreasedEnergyShieldPercentAndStunRecovery3'))
    .addMod(mods.fromId('FireResist1'))
    .addMod(mods.fromId('ReducedLocalAttributeRequirements1'))
    .addMod(mods.fromId('AdditionalMinesPlacedSupportedUber1_'))
    .setProperty('corrupted', true),
];
