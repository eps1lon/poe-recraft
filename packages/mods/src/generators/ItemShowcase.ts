import Item from '../containers/item/Item';
import MasterBench from '../helpers/MasterBench';
import Mod from '../mods/Mod';
import { CraftingBenchOptionsProps, EssenceProps, ModProps } from '../schema';

import Generator, { GeneratorDetails } from './Generator';
import { Alchemy, EnchantmentBench, Essence, Vaal } from './item_orbs';
import { Flags } from '../util/Flags';

/**
 * Masterbench/Currency hybrid
 */
export default class ItemShowcase extends Generator<Mod, Item> {
  public enchantment: EnchantmentBench;
  public master: MasterBench;
  public explicits: Alchemy;
  public vaal: Vaal;

  private essences: EssenceProps[];

  constructor(
    props: ModProps[],
    options: CraftingBenchOptionsProps[],
    essences: EssenceProps[],
  ) {
    const enchantment = EnchantmentBench.build(props);
    const master = MasterBench.build(options);
    const explicits = Alchemy.build(props);
    const vaal = Vaal.build(props);

    const mods = [
      ...enchantment.mods,
      ...explicits.mods,
      ...vaal.mods,
      ...master.getAvailableMods(),
    ];

    super(mods);

    this.enchantment = enchantment;
    this.master = master;
    this.explicits = explicits;
    this.vaal = vaal;

    this.essences = essences;
  }

  /**
   * only abstract showcase, not for actual usage
   */
  public applyTo(item: Item) {
    return item;
  }

  /**
   * not applicable to anything
   * @param item
   */
  public applicableTo(item: Item): Flags {
    return {
      not_applicable: true,
    } as Flags;
  }

  /**
   * greps mod::applicableTo and (if implemented) mod::spawnableOn
   * if we have all the space for mods we need
   *
   * @returns master-, enchantment-, vaal-, explicit-, (top tier) essence-mods
   */
  public modsFor(
    item: Item,
    whitelist: string[] = [],
  ): Array<GeneratorDetails<Mod>> {
    const details = [
      ...this.master.modsFor(item, whitelist),
      ...this.enchantment.modsFor(item, whitelist),
      ...this.explicits.modsFor(item, whitelist),
      ...this.vaal.modsFor(item, whitelist),
      ...this.essences
        .map(props => {
          const essence = new Essence(props, []);
          if (essence.isTopTier()) {
            return essence.chooseMod(item);
          } else {
            return undefined;
          }
        })
        .filter(
          (mod): mod is Mod => {
            return mod !== undefined;
          },
        )
        .map(mod => ({ mod })),
    ];

    // flow cant merge object types
    // { mod: OneMod, prop: number } | { mod: AnotherMod, prop: number }
    // will not become { mod: OneMod | AnotherMod, prop: number }
    // and for some reason ['a', 'b'] cant be cast to Array<number | string>
    return details;
  }
}
