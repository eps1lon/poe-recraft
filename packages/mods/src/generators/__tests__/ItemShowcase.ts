import { createTables } from '../../__fixtures__/util';

import ItemShowcase from '../ItemShowcase';

const { craftingbenchoptions, essences, items, mods } = createTables();

it('should build', () => {
  const showcase = new ItemShowcase(
    mods.all(),
    craftingbenchoptions.all(),
    essences.all(),
  );

  expect(showcase).toBeInstanceOf(ItemShowcase);
  expect(showcase.mods.length).toBeGreaterThan(0);
});

it('should not do anything, just showcase', () => {
  const showcase = new ItemShowcase(
    mods.all(),
    craftingbenchoptions.all(),
    essences.all(),
  );
  const greaves = items.fromName('Iron Greaves');

  expect(showcase.applyTo(greaves)).toBe(greaves);
  expect(showcase.modsFor(greaves).length).toBeGreaterThan(0);
});

it('should use special rarity', () => {
  const showcase = new ItemShowcase(
    mods.all(),
    craftingbenchoptions.all(),
    essences.all(),
  );

  const life = mods.fromId('IncreasedLife0');
  const greaves = items.fromName('Iron Greaves').addMod(life);

  // although this item is white and already has a prefix the showcase
  // considers this item to be rare
  expect(
    showcase
      .modsFor(greaves)
      .filter(({ mod }) => mod.isPrefix() && !mod.isMasterMod())
      .map(({ mod }) => mod.props.id),
  ).toMatchSnapshot();
});

it('is not applicable to anything', () => {
  const showcase = new ItemShowcase(
    mods.all(),
    craftingbenchoptions.all(),
    essences.all(),
  );

  expect(showcase.isApplicableTo(items.fromName('Iron Greaves'))).toBe(false);
});
