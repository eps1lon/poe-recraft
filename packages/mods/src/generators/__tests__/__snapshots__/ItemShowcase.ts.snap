// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should use special rarity 1`] = `
Array [
  "LocalIncreasedPhysicalDamageReductionRatingPercent1",
  "LocalIncreasedPhysicalDamageReductionRatingPercent2",
  "LocalIncreasedPhysicalDamageReductionRatingPercent3",
  "LocalIncreasedPhysicalDamageReductionRatingPercent4",
  "LocalIncreasedPhysicalDamageReductionRatingPercent5",
  "LocalIncreasedPhysicalDamageReductionRatingPercent6",
  "LocalIncreasedPhysicalDamageReductionRatingPercent7",
  "LocalIncreasedPhysicalDamageReductionRatingPercentAndStunRecovery1",
  "LocalIncreasedPhysicalDamageReductionRatingPercentAndStunRecovery2",
  "LocalIncreasedPhysicalDamageReductionRatingPercentAndStunRecovery3",
  "LocalIncreasedPhysicalDamageReductionRatingPercentAndStunRecovery4",
  "LocalIncreasedPhysicalDamageReductionRatingPercentAndStunRecovery5",
  "LocalIncreasedPhysicalDamageReductionRatingPercentAndStunRecovery6",
  "LocalIncreasedPhysicalDamageReductionRating1",
  "LocalIncreasedPhysicalDamageReductionRating2",
  "LocalIncreasedPhysicalDamageReductionRating3",
  "LocalBaseArmourAndLife1",
  "LocalBaseArmourAndLife2",
  "MovementVelocity1",
  "MovementVelocity2",
  "MovementVelocity3",
  "MovementVelocity4",
  "MovementVelocity5",
  "MovementVelocity6",
  "ItemFoundRarityIncreasePrefix1",
  "ItemFoundRarityIncreasePrefix2",
  "LocalIncreasedEnergyShieldEssenceBootsGloves7",
  "IncreasedLifeEssenceBootsGloves1",
  "AttackerTakesDamageEssence7",
  "LocalIncreasedEvasionRatingEssenceGlovesBoots7",
  "MovementVelocityEssence7",
  "IncreasedManaEssence7",
  "LocalIncreasedPhysicalDamageReductionRatingEssenceBootsGloves7",
]
`;
